published/external/xdpapi_helper.h

```diff
@@ -0,0 +1,279 @@
+//
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+//
+
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <xdpapi.h>
+
+#if defined(_KERNEL_MODE)
+
+#include <wdm.h>
+#include <xdp/object.h>
+
+static XDP_API_CLIENT _XdpApiContext = {0};
+static const XDP_API_PROVIDER_DISPATCH *_XdpApi;
+static const XDP_API_PROVIDER_BINDING_CONTEXT *_ProviderBindingContext;
+
+#define XDP_CREATE_HANDLE_WITH_RUNDOWN(FunctionCall, Object, ...)                               \
```

Comment:
why is this a macro - can it be a regular function? nontrivial macros are strongly discouraged, because they are less typesafe and more surprising in subtle behavior.

Reply:
Because each function (XskCreate, XskCreateProgram and XdpInterfaceOpen) takes different number of parameters. Only macro can handle the case.
Which do you prefer? 1 macro or 3 inline functions or directly write this in each function

Reply:
I think doing this pattern in each of the files is acceptable - see XdpIfCreateRxQueue within XDP for an example of similar referencing before creating an object.